zebra-rpc/src/queue.rs

```diff
@@ -0,0 +1,265 @@
+    transactions: HashMap<UnminedTxId, (Arc<Transaction>, Instant)>,
+}
+
+#[derive(Debug)]
+/// The runner
+pub struct Runner {
+    queue: Queue,
+    sender: Sender<Option<UnminedTx>>,
+}
+
+impl Queue {
+    /// Start a new queue
+    pub fn start() -> Runner {
+        let (sender, _receiver) = channel(10);
+
+        let queue = Queue {
+            transactions: HashMap::new(),
+        };
+
+        Runner { queue, sender }
+    }
+
+    /// Get the transactions in the queue.
+    pub fn transactions(&self) -> HashMap<UnminedTxId, (Arc<Transaction>, Instant)> {
+        self.transactions.clone()
+    }
+
+    /// Insert a transaction to the queue.
+    pub fn insert(&mut self, unmined_tx: UnminedTx) {
+        self.transactions
+            .insert(unmined_tx.id, (unmined_tx.transaction, Instant::now()));
+    }
+
+    /// Remove a transaction from the queue.
+    pub fn remove(&mut self, unmined_id: UnminedTxId) {
+        self.transactions.remove(&unmined_id);
+    }
+}
+
+impl Runner {
+    /// Access the sender field of the runner.
+    pub fn sender(&self) -> Sender<Option<UnminedTx>> {
+        self.sender.clone()
+    }
+
+    /// Create a new receiver.
+    pub fn receiver(&self) -> Receiver<Option<UnminedTx>> {
+        self.sender.subscribe()
+    }
+
+    /// Access the queue.
+    pub fn queue(&self) -> Queue {
+        self.queue.clone()
+    }
+
+    /// Get the queue transactions as a `HashSet` of unmined ids.
+    fn transactions_as_hash_set(&self) -> HashSet<UnminedTxId> {
+        let transactions = self.queue.transactions();
+        transactions.iter().map(|t| *t.0).collect()
+    }
+
+    /// Get the queue transactions as a `Vec` of transactions.
+    fn transactions_as_vec(&self) -> Vec<Arc<Transaction>> {
+        let transactions = self.queue.transactions();
+        transactions.iter().map(|t| t.1 .0.clone()).collect()
+    }
+
+    /// Retry sending to memempool if needed.
+    pub async fn run<Mempool, State, Tip>(mut self, mempool: Mempool, state: State, _tip: Tip)
+    where
+        Mempool: Service<Request, Response = Response, Error = BoxError> + Clone + 'static,
+        State: Service<ReadRequest, Response = ReadResponse, Error = zebra_state::BoxError>
+            + Clone
+            + Send
+            + Sync
+            + 'static,
+        Tip: ChainTip + Clone + Send + Sync + 'static,
+    {
+        // TODO: this should be an argument
+        let network = zebra_chain::parameters::Network::Mainnet;
+
+        // TODO: Use tip.best_tip_height()
+        let tip_height = zebra_chain::block::Height(1);
+
+        // get spacing between blocks
+        let spacing = NetworkUpgrade::target_spacing_for_height(network, tip_height);
+
+        let mut receiver = self.sender.subscribe();
+
+        loop {
+            // check the channel for new transactions
+            if let Ok(Some(tx)) = &receiver.recv().await {
+                let _ = &self.queue.insert(tx.clone());
+            }
```

Comment:
Here is an issue that is giving me some problems. The above code will only check for the last transaction in the channel, so this should run in a loop, something as:

```rust
            while let Ok(Some(tx)) = receiver.recv().await { 
                let _ = queue.insert(tx.clone());
            }
```

However, if i do that the program will get blocked in the ` receiver.recv().await` part.

So it has to be spawned (docs actually suggest to spawn: https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html#examples)

To have something as:

```rust
        tokio::spawn(async move {
            while let Ok(Some(tx)) = receiver.recv().await { 
                let _ = queue.insert(tx.clone());
            }
        });
```

However, if we do that, then we start to get troubles with the lifetimes because the references start to be short lived. A similar situation introduced the `Mutex` and the `Arc` before so i am looking for a solution without them.

Reply:
It might help to use [`broadcast::Receiver::try_recv()`](https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Receiver.html#method.try_recv), it will never block.

Reply:
Thanks, this was fixed in https://github.com/ZcashFoundation/zebra/pull/4015/commits/d85a6b662f84e531e654925e04efd856010d3852